### PR TITLE
Wire encryption plugin using chacha cypher

### PR DESCRIPTION
**Pull request by AlexPeshkoff**

### CODE REVIEW

File: builds/posix/Makefile.in [hunk 1]
```diff
@@ -565,6 +566,11 @@ $(LEGACY_USER_MANAGER):	$(LEGACY_USERS_MANAGE_Objects) $(COMMON_LIB)
 	$(LINK_PLUGIN) $(call LIB_LINK_SONAME,$(notdir $@).0) -o $@ $^ $(LINK_PLUG_LIBS) $(FIREBIRD_LIBRARY_LINK)\
  $(call LIB_LINK_DARWIN_INSTALL_NAME,plugins/libLegacy_UserManager.$(SHRLIB_EXT))
 
+chacha:	$(CHACHA)
+
+$(CHACHA):	$(Chacha_Objects) $(COMMON_LIB)
+	$(LINK_PLUGIN) $(call LIB_LINK_SONAME,$(notdir $@).0) -o $@ $^ $(LINK_PLUG_LIBS) $(FIREBIRD_LIBRARY_LINK)
```
Comment (asfernandes) [inline builds/posix/Makefile.in]: Missing LIB_LINK_DARWIN_INSTALL_NAME

File: src/plugins/crypt/chacha/ChaCha.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+class ChaCha FB_FINAL : public StdPlugin<IWireCryptPluginImpl<ChaCha, CheckStatusWrapper> >
+{
+public:
+	explicit ChaCha(IPluginConfig*)
+		: en(NULL), de(NULL), iv(getPool())
+	{ }
+
+	~ChaCha()
+	{
+		delete en;
```
Comment (asfernandes) [inline src/plugins/crypt/chacha/ChaCha.cpp]: Whu not use AutoPtr?

**Comment by aafemt** [comment]
With this plugin do Firebird allow using wire encryption with legacy and sspi authentification?

**Comment by AlexPeshkoff** [comment]
On 2019-12-24 13:36, Dimitry Sibiryakov wrote:
>
> With this plugin do Firebird allow using wire encryption with legacy
>

No. Chacha is less critical to key quality but I see no way to have a 
session key with DES.

> and sspi authentification?
>

SSPI has many implementations. If windows authentication is using active 
directory / kerberos even 'good old' RC4 can be used. With NTLM session 
key obviously has very poor quality and RC4 is bad for that case. Chacha 
in that case is certainly better but I can't say for sure is it good 
enough or not.

PS. The primary goal of adding this plugin is to give an answer to the 
people who cry 'RC4 is bad-bad-bad!' not understanding that's it's bad 
only in _some_ cases. Well - instead of trying to explain something it's 
simpler to add what they want.

**Comment by hvlad** [comment]
@AlexPeshkoff : could you provide minimal documentation about it ? At least at firebird.conf.

**Comment by AlexPeshkoff** [comment]
On 2019-12-24 20:12, Vlad Khorsun wrote:
>
> @AlexPeshkoff <https://github.com/AlexPeshkoff> : could you provide 
> minimal documentation about it ? At least at firebird.conf.
>

Я планирую там доработать кроме документации ещё кое-что но для 
успокоения души прошу проверить - свяжется ли c 4 у которой вот такая 
строка в firebird.conf:

WireCryptPlugin = ChaCha, Arc4

стандартный клиент от 3 на винде.

A.

**Pull request merged**